File: backend/services/weather_service.py
import json
import urllib.request
import urllib.error
import time
import logging
from datetime import datetime
import urllib.parse
from config import Config

logger = logging.getLogger(__name__)

class WeatherService:
    def __init__(self):
        self.lat = Config.WEATHER_LAT
        self.lon = Config.WEATHER_LON
        self.timezone = Config.WEATHER_TIMEZONE
        self.cache_ttl = Config.WEATHER_CACHE_TTL
        self.cache = {
            'data': None,
            'timestamp': 0
        }
        
        self.weather_conditions = {
            0: "Clear sky", 1: "Mainly clear", 2: "Partly cloudy", 3: "Overcast",
            45: "Foggy", 48: "Depositing rime fog",
            51: "Light drizzle", 53: "Moderate drizzle", 55: "Dense drizzle",
            61: "Slight rain", 63: "Moderate rain", 65: "Heavy rain",
            71: "Slight snow", 73: "Moderate snow", 75: "Heavy snow",
            80: "Slight rain showers", 81: "Moderate rain showers", 82: "Violent rain showers",
            95: "Thunderstorm", 96: "Thunderstorm with slight hail", 99: "Thunderstorm with heavy hail"
        }

    def _get_condition(self, code):
        return self.weather_conditions.get(code, "Unknown")

    def _calculate_weather_factor(self, code):
        if code in [63, 81, 73]: # Moderate rain/snow
            return 1.10
        elif code in [65, 82, 95, 96, 99, 75]: # Heavy rain, thunderstorms, heavy snow
            return 1.20
        elif code in [45, 48]: # Fog
            return 1.05
        return 1.0

    def get_weather(self):
        current_time = time.time()
        
        if self.cache['data'] and (current_time - self.cache['timestamp']) < self.cache_ttl:
            logger.info("Returning cached weather data")
            data = dict(self.cache['data'])
            data['cached'] = True
            return data
            
        params = {
            'latitude': self.lat,
            'longitude': self.lon,
            'current': 'temperature_2m,relative_humidity_2m,precipitation,rain,wind_speed_10m,weather_code',
            'hourly': 'temperature_2m,precipitation_probability,precipitation,rain,wind_speed_10m,weather_code',
            'timezone': self.timezone
        }
        
        query_string = urllib.parse.urlencode(params)
        url = f"https://api.open-meteo.com/v1/forecast?{query_string}"
        
        try:
            req = urllib.request.Request(url, headers={'User-Agent': 'PRAVAAH/1.0'})
            with urllib.request.urlopen(req) as response:
                raw_data = json.loads(response.read().decode('utf-8'))
                
            current = raw_data.get('current', {})
            hourly = raw_data.get('hourly', {})
            
            weather_code = current.get('weather_code', 0)
            
            hourly_data = []
            if hourly:
                for i in range(min(24, len(hourly.get('time', [])))):
                    hourly_data.append({
                        'time': hourly['time'][i],
                        'temperature': hourly['temperature_2m'][i],
                        'rain_probability': hourly['precipitation_probability'][i],
                        'rain': hourly['rain'][i],
                        'wind_speed': hourly['wind_speed_10m'][i],
                        'condition': self._get_condition(hourly['weather_code'][i]),
                        'weather_code': hourly['weather_code'][i]
                    })
                    
            precip_prob = hourly_data[0]['rain_probability'] if hourly_data else 0

            formatted_data = {
                "location": {
                    "name": "Mumbai",
                    "latitude": self.lat,
                    "longitude": self.lon,
                    "timezone": self.timezone
                },
                "current": {
                    "temperature": current.get('temperature_2m', 0.0),
                    "humidity": current.get('relative_humidity_2m', 0),
                    "rain": current.get('rain', 0.0),
                    "precipitation_probability": precip_prob,
                    "wind_speed": current.get('wind_speed_10m', 0.0),
                    "weather_code": weather_code,
                    "condition": self._get_condition(weather_code)
                },
                "hourly": hourly_data,
                "weather_factor": self._calculate_weather_factor(weather_code),
                "source": "Open-Meteo",
                "cached": False,
                "timestamp": datetime.now().isoformat()
            }
            
            self.cache['data'] = formatted_data
            self.cache['timestamp'] = current_time
            return formatted_data
            
        except Exception as e:
            logger.error(f"Failed to fetch weather data: {str(e)}")
            if self.cache['data']:
                logger.info("Returning stale cached weather data due to error")
                data = dict(self.cache['data'])
                data['cached'] = True
                return data
            return None

weather_service = WeatherService()

File: frontend/src/services/weatherService.js
/**
 * PRAVAAH Weather Service
 * Integrates Open-Meteo real-time weather data for Mumbai
 */

import api from '../lib/api'

const OPEN_METEO_URL = 'https://api.open-meteo.com/v1/forecast'
const MUMBAI_LAT = 19.0760
const MUMBAI_LON = 72.8777

const WEATHER_CODES = {
  0: { condition: 'Clear sky', icon: '☀️' },
  1: { condition: 'Mainly clear', icon: '🌤️' },
  2: { condition: 'Partly cloudy', icon: '⛅' },
  3: { condition: 'Overcast', icon: '☁️' },
  45: { condition: 'Foggy', icon: '🌫️' },
  48: { condition: 'Rime fog', icon: '🌫️' },
  51: { condition: 'Light drizzle', icon: '🌦️' },
  53: { condition: 'Drizzle', icon: '🌦️' },
  55: { condition: 'Dense drizzle', icon: '🌧️' },
  61: { condition: 'Slight rain', icon: '🌧️' },
  63: { condition: 'Moderate rain', icon: '🌧️' },
  65: { condition: 'Heavy rain', icon: '🌧️' },
  71: { condition: 'Slight snow', icon: '❄️' },
  73: { condition: 'Moderate snow', icon: '❄️' },
  75: { condition: 'Heavy snow', icon: '❄️' },
  80: { condition: 'Rain showers', icon: '🌦️' },
  81: { condition: 'Moderate showers', icon: '🌧️' },
  82: { condition: 'Violent showers', icon: '⛈️' },
  95: { condition: 'Thunderstorm', icon: '⛈️' },
  96: { condition: 'Thunderstorm + hail', icon: '⛈️' },
  99: { condition: 'Severe thunderstorm', icon: '⛈️' }
}

function getWeatherInfo(code) {
  return WEATHER_CODES[code] || { condition: 'Unknown', icon: '🌡️' }
}

function calculateWeatherFactor(weatherCode, rainMm) {
  if (weatherCode >= 95) return 1.25 // Thunderstorm
  if (weatherCode >= 65 || rainMm > 7) return 1.20 // Heavy rain
  if (weatherCode >= 61 || rainMm > 2.5) return 1.15 // Moderate rain
  if (weatherCode >= 51 || rainMm > 0) return 1.10 // Light rain/drizzle
  if (weatherCode >= 45) return 1.05 // Fog
  return 1.0 // Clear/cloudy
}

let weatherCache = null
let cacheTimestamp = 0
const CACHE_TTL = 10 * 60 * 1000 // 10 minutes

/**
 * Fetch weather data.
 * First tries the backend /api/weather endpoint.
 * Falls back to calling Open-Meteo directly from frontend (it supports CORS).
 */
export async function getWeather() {
  // Return cached if fresh
  if (weatherCache && (Date.now() - cacheTimestamp) < CACHE_TTL) {
    return { ...weatherCache, cached: true }
  }

  // Try backend first
  try {
    const res = await api.get('/weather')
    weatherCache = res.data
    cacheTimestamp = Date.now()
    return weatherCache
  } catch (_) {
    // Backend unavailable, try direct Open-Meteo
  }

  // Direct Open-Meteo call (CORS supported)
  try {
    const params = new URLSearchParams({
      latitude: MUMBAI_LAT,
      longitude: MUMBAI_LON,
      current: 'temperature_2m,relative_humidity_2m,precipitation,rain,wind_speed_10m,weather_code',
      hourly: 'temperature_2m,precipitation_probability,precipitation,rain,wind_speed_10m,weather_code',
      timezone: 'Asia/Kolkata',
      forecast_days: 1
    })

    const response = await fetch(`${OPEN_METEO_URL}?${params}`)
    if (!response.ok) throw new Error(`Open-Meteo HTTP ${response.status}`)
    const raw = await response.json()

    const currentCode = raw.current?.weather_code ?? 0
    const currentInfo = getWeatherInfo(currentCode)
    const currentRain = raw.current?.rain ?? 0

    // Find current hour index
    const now = new Date()
    const currentHour = now.getHours()
    const hourlyTimes = raw.hourly?.time || []

    // Get next 6 hours of forecast
    const hourlyForecasts = []
    for (let i = 0; i < hourlyTimes.length && hourlyForecasts.length < 6; i++) {
      const forecastHour = new Date(hourlyTimes[i]).getHours()
      const forecastDate = new Date(hourlyTimes[i])
      if (forecastDate >= now || forecastHour >= currentHour) {
        const code = raw.hourly.weather_code?.[i] ?? 0
        hourlyForecasts.push({
          time: hourlyTimes[i],
          hour: `${String(forecastHour).padStart(2, '0')}:00`,
          temperature: raw.hourly.temperature_2m?.[i] ?? 0,
          rain_probability: raw.hourly.precipitation_probability?.[i] ?? 0,
          rain: raw.hourly.rain?.[i] ?? 0,
          wind_speed: raw.hourly.wind_speed_10m?.[i] ?? 0,
          weather_code: code,
          condition: getWeatherInfo(code).condition,
          icon: getWeatherInfo(code).icon
        })
      }
    }

    const weather = {
      location: {
        name: 'Mumbai',
        latitude: MUMBAI_LAT,
        longitude: MUMBAI_LON,
        timezone: 'Asia/Kolkata'
      },
      current: {
        temperature: raw.current?.temperature_2m ?? 0,
        humidity: raw.current?.relative_humidity_2m ?? 0,
        rain: currentRain,
        precipitation: raw.current?.precipitation ?? 0,
        wind_speed: raw.current?.wind_speed_10m ?? 0,
        weather_code: currentCode,
        condition: currentInfo.condition,
        icon: currentInfo.icon
      },
      hourly: hourlyForecasts,
      weather_factor: calculateWeatherFactor(currentCode, currentRain),
      source: 'Open-Meteo',
      cached: false,
      timestamp: new Date().toISOString()
    }

    weatherCache = weather
    cacheTimestamp = Date.now()
    return weather
  } catch (err) {
    console.error('[PRAVAAH] Weather fetch failed:', err)
    // Return cached if available, null otherwise
    if (weatherCache) return { ...weatherCache, cached: true }
    return null
  }
}

export { getWeatherInfo, calculateWeatherFactor, WEATHER_CODES }
